_source/_posts/2020-06-16-security-audits.adoc

```diff
@@ -0,0 +1,112 @@
+image: blog/security-audits/security-audits-social.png
+type: awareness
+---
+:page-liquid:
+
+I've always had mixed feelings about security reviews. My opinions range from "It's a waste of time" to "I'd really like someone to verify my code." This likely stems from the fact that various shops I've worked at have all handled them differently—from no process at all, to grepping source for keywords like "password," to detailed code analysis.
+
+Like everything in the security world, the answer isn't a yes or no, it's a spectrum that is different for every project. If you are writing a password hashing algorithm, you have different requirements than a logging framework. Same with an online banking website vs. a command-line tool. In this post, I'll walk through a recent security audit performed on a library I've been working on: https://github.com/paseto-toolkit/jpaseto[JPaseto], a library that creates and parses security tokens.
+
+image::{% asset_path 'blog/security-audits/security-audits-social.png' %}[alt=Do you need a security audit?,width=800,align=center]
+
+== Different Types of Security Reviews
+
+As I mentioned before, there are a variety of ways you could inspect our codebase. Regardless of your needs, automating your process is critical. If your process for security reviews involves running a few commands and looking at the results, it can probably be automated. Even better, include it as part of your build and CI process.
+
+In the Java world, I'm a big fan of https://spotbugs.github.io/[SpotBugs] and https://find-sec-bugs.github.io/[Find Security Bugs]. No tool is going to find all bugs, but these tools have found more issues than I would have expected and more importantly, have taught me a great deal about various types of bugs and how to avoid them.
+
+In-depth code analysis is great, but when it comes from outside your team, they are generally slow and too costly (time and money) to do against every commit. This often leaves teams to do them "as needed." The catch 22 is—if you are not a security expert—how do you know when they are needed? One option is to schedule them periodically—quarterly, for example—or some other frequency that matches the pace of your development.
+
+== Outside Audit
+
+For the first release of https://github.com/paseto-toolkit/jpaseto[JPaseto] we hired https://paragonie.com/[Paragon Initiative Enterprises] to do a security audit. They had domain expertise, as they helped author the https://paseto.io/rfc/[PASETO RFC]. The turn around was fast, and they came back with a few issues:
+
+. Constant Time Comparison
++
+This was the most critical of the issues they found. The library was making a comparison of a hashed value using Java's `Arrays.equals()` method which fails fast, because it's optimized for speed. The code should have been using a time constant comparison such as `MessageDigest.isEqual()` to prevent timing attacks.
```

Comment:
I just learned about the + line in AsciiDoc a few weeks ago! It's pretty slick and can be used to add code blocks into lists too.

Reply:
Nice! I didn't know that!